clang/lib/Driver/Driver.cpp

```diff
@@ -2886,6 +2886,9 @@ class OffloadingActionBuilder final {
     /// Append linker actions generated by the builder.
     virtual void appendLinkDependences(OffloadAction::DeviceDependences &DA) {}
 
+    /// Append linker actions generated by the builder.
+    virtual void addDeviceLinkDependencies(OffloadDepsJobAction *DA) {}
```

Comment:
```suggestion
    virtual void addDeviceLinkDependencies(OffloadDepsJobAction *DA) = 0;
```

Reply:
That would make this class abstract, but I did not want to do it because it would force successors which we do not want to change (CudaActionBuilder, HIPActionBuilder) to define this method. I would rather keep a default implementation which does nothing.